pkg/csi/service/wcp/controller.go

```diff
@@ -1278,28 +1279,24 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
 							"failed to connect to Virtual Center: %s", vc.Config.Host)
 					}
-					podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
-					if err != nil {
-						if err == cnsvsphere.ErrVMNotFound {
-							log.Infof("virtual machine not found for vmUUID %q. "+
-								"Thus, assuming the volume is detached.", v)
-							break
-						}
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"failed to get the PodVM Moref from the PodVM UUID: %s in datacenter: %s with err: %+v",
-							v, dcMorefValue, err)
-					}
 					isStillAttached := false
 					timeout := 4 * time.Minute
 					pollTime := time.Duration(5) * time.Second
-					err = wait.Poll(pollTime, timeout, func() (bool, error) {
-						podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
+					var podVM *cnsvsphere.VirtualMachine
+					err = wait.PollImmediate(pollTime, timeout, func() (bool, error) {
+						podVM, err = getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
 						if err != nil {
 							if err == cnsvsphere.ErrVMNotFound {
 								log.Infof("virtual machine not found for vmUUID %q. "+
 									"Thus, assuming the volume is detached.", v)
 								return true, err
 							}
+							if err == cnsvsphere.ErrInvalidVC {
+								log.Errorf("failed to get the PodVM Moref from the PodVM UUID: %s in datacenter: "+
+									"%s with err: %+v", v, dcMorefValue, err)
+								return true, err
```

Comment:
If VC is invalid, shouldn't we return `fail, err`?

Reply:
Got on a call with Divyen and decided to return `false, err`

Reply:
> // PollImmediate tries a condition func until it returns true, an error, or the timeout
> // is reached.

Based on the docs either `false, err` or `true, err` would work the same way here, right?

Reply:
yes. `false` will just indicate here that condition we were looking still not done.